libraries/ms-common/src/stm32f0xx/flash.c

```diff
@@ -0,0 +1,60 @@
+#include "flash.h"
+#include <string.h>
+#include "stm32f0xx.h"
+
+StatusCode flash_init(void) {
+  // Flash prefetch and latency are set up in system init
+  return STATUS_CODE_OK;
+}
+
+StatusCode flash_read(uintptr_t address, size_t read_bytes, uint8_t *buffer, size_t buffer_len) {
+  if (buffer_len < read_bytes || address < FLASH_BASE_ADDR ||
+      (address + read_bytes) > FLASH_END_ADDR) {
+    return status_code(STATUS_CODE_OUT_OF_RANGE);
+  }
+
+  // may want to move all flash-related functions into RAM so we don't block the CPU
+  memcpy(buffer, (void *)address, read_bytes);
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode flash_write(uintptr_t address, uint8_t *buffer, size_t buffer_len) {
+  if (address < FLASH_BASE_ADDR || (address + buffer_len) > FLASH_END_ADDR) {
+    return status_code(STATUS_CODE_OUT_OF_RANGE);
+  } else if (buffer_len % FLASH_WRITE_BYTES != 0 || address % FLASH_WRITE_BYTES != 0) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  uint32_t *data = (uint32_t *)buffer;
+  size_t data_len = buffer_len / FLASH_WRITE_BYTES;
+
+  FLASH_Unlock();
```

Comment:
Good question - the bus stalls if an attempt to read is made during a erase/write, but the read operation completes successfully once the previous operation has completed. I suppose there's a possibility of an interrupt attempting to read from flash (code fetch) while an erase/write is occurring, so the code fetch will stall until the operation is complete. Then, the interrupt handler will run to completion and then return back to our flash function.

I don't think this will ever cause an error, but I can wrap it in a critical section just for safety.